util/rateLimit.go

```diff
@@ -0,0 +1,461 @@
+	erl.clientLock.Lock()
+	if _, exists := erl.capacityByClient[c]; exists {
+		return capacityQueue(nil), fmt.Errorf("client already has a reservation")
+	}
+	// guard against overprovisioning, if there is less than a reservedCapacity amount left
+	remaining := erl.MaxCapacity - (erl.CapacityPerReservation * len(erl.capacityByClient))
+	if erl.CapacityPerReservation > remaining {
+		return capacityQueue(nil), fmt.Errorf("not enough capacity to reserve for client: %d remaining, %d requested", remaining, erl.CapacityPerReservation)
+	}
+	// make capacity for the provided client
+	q := capacityQueue(make(chan capacity, erl.CapacityPerReservation))
+	erl.capacityByClient[c] = q
+	erl.clientLock.Unlock()
+
+	// fill the reservation with capacity optimistically before consuming the shared capacity
+	for i := 0; i < erl.CapacityPerReservation; i++ {
+		q <- capacity{}
+	}
+	// create a thread to drain the capacity from sharedCapacity in a blocking way
+	go func() {
+		for i := 0; i < erl.CapacityPerReservation; i++ {
+			erl.sharedCapacity.blockingConsume()
+		}
+	}()
+	return q, nil
+}
+
+// closeReservation will remove the client mapping to capacity channel,
+// and will kick off a routine to drain the capacity and replace it to the shared capacity
+func (erl ElasticRateLimiter) closeReservation(c ErlClient) {
+	erl.clientLock.Lock()
+	q, exists := erl.capacityByClient[c]
+	// guard clauses, and preventing the ElasticRateLimiter from draining its own sharedCapacity
+	if !exists || q == erl.sharedCapacity {
+		return
+	}
+	delete(erl.capacityByClient, c)
+	erl.clientLock.Unlock()
+
+	// start a routine to consume capacity from the closed reservation, and return it to the sharedCapacity
+	go func() {
+		for i := 0; i < erl.CapacityPerReservation; i++ {
+			erl.sharedCapacity.blockingRelease()
+			q.blockingConsume()
+		}
+	}()
+}
+
+// CongestionManager is an interface for tracking events which happen to capacityQueues
+type CongestionManager interface {
+	Start(ctx context.Context, wg *sync.WaitGroup)
+	Consumed(c ErlClient, t time.Time)
+	Served(t time.Time)
+	ShouldDrop(c ErlClient) bool
+}
+
+type event struct {
+	c ErlClient
+	t time.Time
+}
+
+type shouldDropQuery struct {
+	c   ErlClient
+	ret chan bool
+}
+
+// "Random Early Detection" congestion manager,
+// will propose to drop messages proportional to the caller's request rate vs Average Service Rate
+type redCongestionManager struct {
+	runLock                *deadlock.Mutex
+	running                bool
+	window                 time.Duration
+	consumed               chan event
+	served                 chan event
+	shouldDropQueries      chan shouldDropQuery
+	targetRate             float64
+	targetRateRefreshTicks int
+	// consumed is the only value tracked by-queue. The others are calculated in-total
+	// TODO: If we desire later, we can add mappings onto release/done for more insight
+	consumedByClient map[ErlClient]*[]time.Time
```

Comment:
this is not needed, and its never really used.

Reply:
I wish Go had a `#[cfg(test)]` flag on things.
Leave a comment saying it's just for inspection during tests.